### PR TITLE
fix: slashed fp and BTC rewards tracker

### PR DESCRIPTION
**Pull request by RafilxTenfen**
`EventPowerDistUpdate_SlashedFp` and [`RewardBTCStaking`](https://github.com/babylonlabs-io/babylon/blob/b132d16483d2e12ac68be74a524a4b7121edac65/x/incentive/keeper/btc_staking_gauge.go#L19) are called at different blocks. 
The rewards are being called a few blocks behind and if the slashed fp is pruned in BTC reward tracker data, it fails to find and panics 

On babylond-deployments an fp is being slashed at block height 46, the call of `RewardBTCStaking` happens at Babylon block 48 but is called to distribute rewards of voting power table from block 44 (which didn't had the FP slashed yet) which then checks for fp rewards for the FP that was pruned at the slashing processed event.

```
6:04PM INF finalizing commit of block hash=153E869F7B0056A9CD8208301F2019FFACA2BEFBBB35AEADF70DCA14C15F3972 height=47 module=consensus num_txs=2 root=62B5F025459DEEBE2EE6F5D869C5BCBE903299FF80825FB9DE62551E6B725D52
 ERR CONSENSUS FAILURE!!! err="failed to add fp rewards for btc delegation bbn1wwkaq6z3kdkekm7ltwxng4ftvzux8gzp2xjx8d at height 44: finality provider current rewards not found
``` 

```shell
$~ babylond q btcstaking finality-providers -o json | jq
...
    {
      "description": {
        "moniker": "Finality Provider 0",
        "identity": "",
        "website": "",
        "security_contact": "",
        "details": ""
      },
      "commission": "0.050000000000000000",
      "addr": "bbn1wwkaq6z3kdkekm7ltwxng4ftvzux8gzp2xjx8d",
      "btc_pk": "6ba387c315dc6105ec02b50684593505103a9f13452d3a597c16ac2f22b924dc",
      "pop": {
        "btc_sig_type": "BIP340",
        "btc_sig": "Q3PcfiKT0fwWOVx95rjV+mQvI/juTbYIV84Hfz2NAmQByiHRXz8OWnFbyugJgOrd/TrqR+6BOgtr10bZUFDFbA=="
      },
      "slashed_babylon_height": "46",
      "slashed_btc_height": 131,
      "height": "46",
      "jailed": false,
      "highest_voted_height": 45
    },
```

Basically, this means we are processing the rewards a few blocks down the road after the slash and then it is not possible to prune the data for reward tracker until that fp slashed height is processed.

The fix for the time being: do nothing at the btc reward tracker structures if some fp gets slashed just continue to store that data. In the future during the issue https://github.com/babylonlabs-io/babylon/issues/362 we can properly handle the pruning of slashed finality provider data

### CODE REVIEW

File: x/finality/keeper/power_dist_change.go [hunk 1]
```diff
@@ -320,6 +323,11 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 	// sort new finality providers in activeBTCDels to ensure determinism
```
Comment (KonradStaniec) [inline x/finality/keeper/power_dist_change.go:323]: comment about a line in which we set `fp.IsSlashed = true`  i.e
```
		if isSlashed {
			fp.IsSlashed = true
			continue
		}
```
shouldn't we also set `fp.TotalBondedSats = 0` there ? this would be inline with previous logic in which we set
`fp.TotalBondedSats = 0` at the begininng and the loop and later do not add any delegations this fp so effectively leaving it with `0`

File: x/finality/keeper/power_dist_change.go [hunk 2]
```diff
@@ -320,6 +323,11 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
...
 	fpActiveBtcPkHexList := make([]string, 0, len(activedSatsByFpBtcPk))
 	for fpBTCPKHex := range activedSatsByFpBtcPk {
+		// if the fp was slashed, should not even be added to the list
+		_, isSlashed := slashedFPs[fpBTCPKHex]
```
Comment (KonradStaniec) [inline x/finality/keeper/power_dist_change.go:327]: hmm I think we should add it the active list and and later replicate the logic from above loop i.e 
```
		if isSlashed {
			fp.IsSlashed = true
			continue
		}
```

as otherwise we won't set jailed flag for new slashed fp. (not sure this is criticial, but this is definitly inconsisent)

File: x/finality/keeper/power_dist_change.go (hunk 1 above)
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:323]: we can, but that would not modify the `newDc` since we are calling `continue` if the fp is slashed

File: x/finality/keeper/power_dist_change.go (hunk 2 above)
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:327]: good point

File: x/finality/keeper/power_dist_change.go [hunk 3]
```diff
@@ -233,10 +233,13 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 			types.EmitSlashedFPEvent(sdkCtx, typedEvent.SlashedFp.Pk)
 			fpBTCPKHex := typedEvent.SlashedFp.Pk.MarshalHex()
 			slashedFPs[fpBTCPKHex] = struct{}{}
-			fp := k.loadFP(ctx, fpByBtcPkHex, fpBTCPKHex)
-			if err := k.IncentiveKeeper.FpSlashed(ctx, fp.Address()); err != nil {
-				panic(err)
-			}
```
Comment (RafilxTenfen) [inline x/finality/keeper/power_dist_change.go:239]: This was the important part

File: test/e2e/btc_rewards_distribution_e2e_test.go [hunk 4]
```diff
@@ -442,7 +445,50 @@ func (s *BtcRewardsDistribution) Test7CheckRewards() {
 	s.NotEmpty(del2DiffRewardsStr)
 }
 
-// TODO(rafilx): Slash a FP and expect rewards to be withdraw.
+// Test8SlashFp slashes the finality provider, but should continue to produce blocks
+func (s *BtcRewardsDistribution) Test8SlashFp() {
```
Comment (RafilxTenfen) [inline test/e2e/btc_rewards_distribution_e2e_test.go:449]: test that verifies by slashing it doesn't panic the rewards anymore

File: x/finality/keeper/power_dist_change.go [hunk 5]
```diff
@@ -332,6 +340,14 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 		newFP := k.loadFP(ctx, fpByBtcPkHex, fpBTCPKHex)
```
Comment (KonradStaniec) [inline x/finality/keeper/power_dist_change.go:340]: I think you are righ about about in above comments (https://github.com/babylonlabs-io/babylon/pull/371/files#r1896052253)

Though one thing that worries me is that we should probably ignore fp here if he is already slashed as even though we do not accept delegations for slashed finality providers, it is still possible to generate voting power events for them:
- it is still possible to activate delegation which was created without inclusion proof
- it is sittl possible to unbond delegation which would create unbonding event

Could you try to add some tests for this cases to verify that it is handled properly currently and add fixe if they dont ? (basically we do not want for slashed fp to ever end up in voting power set ever again)

**Review by KonradStaniec — APPROVED** [review]
Fix looks good 👍 

Added one comment about thing we are possibly missing.

**Review by Lazar955 — APPROVED** [review]

**Review by gitferry — COMMENTED** [review]
Maybe an alternative solution is to add a condition to break the loop if the fp.Slashed in https://github.com/babylonlabs-io/babylon/blob/main/x/finality/types/power_table.go#L77?

**Comment by RafilxTenfen** [comment]
> Maybe an alternative solution is to add a condition to break the loop if the fp.Slashed in https://github.com/babylonlabs-io/babylon/blob/main/x/finality/types/power_table.go#L77?

I totally agree

File: x/finality/keeper/power_dist_change.go (hunk 5 above)
Comment (RafilxTenfen) [reply x/finality/keeper/power_dist_change.go:340]: I am verifying if they are slashed, in the creation of `activedSatsByFpBtcPk` list, but yeah adding the check as Gai [suggested](https://github.com/babylonlabs-io/babylon/pull/371#pullrequestreview-2521909396) seems to make a lot of sense to me
